policy: skip client side validation of policies that can't be validated

oblazek opened the pull request:
Both `cilium policy import` and `cilium policy validate` command not
only use server side validation of policies but both also use client
side validation of policies - meaning they both run `r.Sanitize()`
without being able to properly check cilium-agent's configuration
flags. This can result in being unable to validate/import policy that
uses a field (like fromNodes, toNodes) which is not available by default
and it's configuration is false by default.

In existing examples this "only" runs `r.Sanitize()` twice. The problem
is in a situation where some of the features is not enabled by default.
In that case client side validation fails even though server side
validation would run properly.
This can also be a problem when we do `./check-examples.sh` in
documentation which runs only client side validation.

Let's skip client side validation of rules that are validatable only on
the server side.

```release-note
policy: fix client side validation of policies in policy import/validate command
```

I ran into this problem when in this PR https://github.com/cilium/cilium/pull/31188. 
Related slack thread: https://cilium.slack.com/archives/C2B917YHE/p1709720825271219

joestringer commented on api/v1/openapi.yaml:
```diff
@@ -555,6 +555,7 @@ paths:
       - policy
       parameters:
       - "$ref": "#/parameters/policy-rules"
+      - "$ref": "#/parameters/policy-validate"
```
It doesn't make sense to me to have an API change that modifies validation behaviour. Validation protects the internals of the agent, so should be applied unconditionally server-side. In general, delegating the option to validate completely to clients is not great form since a client that is acting poorly could then endanger the reliability of the server.

oblazek replied:
This is not completely true, the validation actually happens unconditionally server-side anyway.  With the current upstream code it only happens twice (once client-side, once server-side). The API change here is only added here to be able to reuse PUT API call.

joestringer replied:
Maybe to take a step back, could you describe why the client needs to make different API calls to change what the client does? Why wouldn't this just be a CLI flag for the client?

oblazek replied:
you are right, I don't..

joestringer commented on cilium-dbg/cmd/policy_import.go:
```diff
@@ -38,12 +38,6 @@ var policyImportCmd = &cobra.Command{
-				if err := r.Sanitize(); err != nil {
-					Fatalf("%s", err)
-				}
-			}
```
Dropping the sanitization is one option, but I wonder if it would be better to just pass a parameter that can be used to determine how to sanitize the configuration. For instance if you're trying to do some sanitization based on an agent-side setting, then maybe the client by default should not apply that specific validation check, but the server side should validate based on the configuration.

I could imagine passing an object in here that has a function to check the specific option that you want to test. For the server, it will pass an object with a function that checks the state of `pkg/option/...`. For the client, the function would return a result that provides the least strict validation.

oblazek replied:
true enough, see the latest revision please

joestringer commented on cilium-dbg/cmd/policy_validate.go:
```diff
@@ -22,22 +22,17 @@ var policyValidateCmd = &cobra.Command{
+				Fatalf("Cannot marshal policy %s: %s\n", path, err)
 			}
-			if policyVerbose {
+			if resp, err := client.PolicyValidate(string(jsonPolicy)); err != nil {
```
I suspect the reason CI is failing is because of the swap from `Sanitize()` -> `PolicyValidate()` here, introducing a hard dependency on the server.

oblazek replied:
yeah true, it's not a full server anyway

joestringer commented on pkg/policy/api/rule_validation.go:
```diff
@@ -113,84 +115,88 @@ func (i *IngressRule) sanitize() error {
 
 	if len(i.FromNodes) > 0 && !option.Config.EnableNodeSelectorLabels {
-		return fmt.Errorf("FromNodes rules can only be applied when the %q flag is set", option.EnableNodeSelectorLabels)
+		if !softValidation {
```
nit: I would suggest to use the following form for boolean checks:

```
if foo {
...
} else {
...
}
```

If the form is instead `if !foo { ... } else { ... }`, then it makes it more difficult to understand the `else` case when reading the code.

(Same feedback elsewhere in this PR)

joestringer commented on pkg/policy/api/rule_validation.go:
```diff
@@ -113,84 +115,88 @@ func (i *IngressRule) sanitize() error {
+		if !softValidation {
+			return fmt.Errorf("FromNodes rules can only be applied when the %q flag is set", option.EnableNodeSelectorLabels), false
+		} else {
+			return fmt.Errorf("FromNodes rules can only be applied when the %q flag is set", option.EnableNodeSelectorLabels), true
```
Hmm. The downside of this approach is that this will shortcut all of the remaining checks below.

Consider for the case where the user specifies this new flag on the CLI commandline, then loads a policy with two possible issues:
- The policy uses node selector labels, which may be OK if the server (cilium-agent) has a flag enabled
- The policy has an invalid `FromEndpoints` field.

In this scenario, the client will pass the flag and the client-side validation will exit early on this line, without checking `FromEndpoints`. Then the client will pass the same policy to the server, which will later fail on the `FromEndpoints` validation.

What do you think the behaviour should be in this case? Since the server still does the validation, I think that the behaviour will be correct (deny the invalid policy). In some ways it would be nicer if the client could validate the entire policy before sending it to the server.

joestringer replied:
One option would be to declare this error statically at the top of the file:

```
errFromNodesRequiresNodeSelectorOption = fmt.Errorf("FromNodes rules can only be applied when the %q flag is set", option.EnableNodeSelectorLabels)
```

then,
```suggestion
			retErr = errFromNodesRequiresNodeSelectorOption
```

and continue to the end of the function. If any other error occurs, it would immediately return that "fatal" error. However, if the control reaches the end of the function, it could just return this `retErr` to the caller.

From the caller, it could check `if errors.Is(err, api.errFromNodesRequiresNodeSelectorOption) { // Don't error out }`.

In this case, we might not even need the new skipped boolean return value.

oblazek replied:
yeah, this is so much better this way :+1:

oblazek committed:
policy: skip validation of policies behind a flag

Both `cilium policy import` and `cilium policy validate` command not
only use server side validation of policies but both also use client
side validation of policies - meaning they both run `r.Sanitize()`
without being able to properly check cilium-agent's configuration
flags. This can result in being unable to validate/import policy that
uses a field (like fromNodes, toNodes) which is not available by default
and it's configuration is false by default.

In existing examples this "only" runs `r.Sanitize()` twice. The problem
is in a situation where some of the features are not enabled by default.
In that case client side validation fails even though server side
validation would run properly.
This can also be a problem when we do `./check-examples.sh` in
documentation which runs only client side validation.

Let's skip client side validation of rules that are validatable only on
the server side.

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

learnitall reviewed (approved):
Thank you!

joestringer commented on cilium-dbg/cmd/policy_import.go:
```diff
@@ -40,7 +42,12 @@ var policyImportCmd = &cobra.Command{
-					Fatalf("%s", err)
+					if errors.Is(err, api.ErrFromToNodesRequiresNodeSelectorOption) {
+						// Don't error out as this can't be validated client-side
+						fmt.Printf("Validation of policy %s has been skipped as it can't be validated.\n", path)
```
nit: I think the current wording of this (and the other similar message) implies that no validation occurs, but from a user's perspective, that's not entirely true. I think that a message like this would be more clear to the user:

```suggestion
						fmt.Printf("Validation of policy %s has been skipped in the client, further validation will occur server-side.\n", path)
```

joestringer reviewed (approved):
Oh yeah, that last suggestion made the change much tidier :tada: . I have one minor comment about the wording for the user-facing message, but otherwise the change looks good.

joestringer commented on cilium-dbg/cmd/policy_validate.go:
```diff
@@ -22,12 +25,19 @@ var policyValidateCmd = &cobra.Command{
+					valid = false
+					if errors.Is(err, api.ErrFromToNodesRequiresNodeSelectorOption) {
+						// Don't error out as this can't be validated client-side
+						fmt.Printf("Validation of policy %s has been skipped as it can't be validated.\n", path)
```
nit: This should use the same format string as the other error.

oblazek replied:
yeah, my bad.. fixed

joestringer reviewed (approved):
LGTM thanks!
